Add device MAC settings configuration to the Console

bafonins opened the pull request:
<!--
Thanks for submitting a pull request. Please fill the template below,
otherwise we will not be able to process this pull request.
-->

#### Summary
<!--
A short summary, referencing related issues:
Closes #0000, References #0000, etc.
-->

References: https://github.com/TheThingsNetwork/lorawan-stack/issues/2047
This PR adds high priority fields from `mac_settings`

<details>
<summary>Screenshots&Details</summary>

ABP:

<img width="876" alt="Screenshot 2020-04-02 at 14 03 27" src="https://user-images.githubusercontent.com/16374166/78243960-65ab0780-74ed-11ea-8469-08957ad6a519.png">

OTAA:

<img width="876" alt="Screenshot 2020-04-02 at 14 03 43" src="https://user-images.githubusercontent.com/16374166/78243974-68a5f800-74ed-11ea-85dc-ffe06978efbd.png">

Multicast:

<img width="876" alt="Screenshot 2020-04-02 at 14 03 53" src="https://user-images.githubusercontent.com/16374166/78243978-693e8e80-74ed-11ea-9da8-62bd3b867482.png">

Class B:

<img width="857" alt="Screenshot 2020-04-02 at 14 42 33" src="https://user-images.githubusercontent.com/16374166/78245560-5a0d1000-74f0-11ea-88c4-80af36c6a6be.png">



Values presented to the user:
<img width="828" alt="Screenshot 2020-03-31 at 17 17 34" src="https://user-images.githubusercontent.com/16374166/78134579-11d6ea80-7429-11ea-9d5d-5e595425992f.png">
<img width="828" alt="Screenshot 2020-03-31 at 17 17 42" src="https://user-images.githubusercontent.com/16374166/78134581-126f8100-7429-11ea-8e51-92a6370a4ab5.png">
<img width="828" alt="Screenshot 2020-03-31 at 17 17 57" src="https://user-images.githubusercontent.com/16374166/78134584-13081780-7429-11ea-9a4c-d8e604fbff08.png">
<img width="827" alt="Screenshot 2020-03-31 at 17 18 04" src="https://user-images.githubusercontent.com/16374166/78134586-13a0ae00-7429-11ea-94db-a77b99701359.png">
<img width="876" alt="Screenshot 2020-04-02 at 14 22 04" src="https://user-images.githubusercontent.com/16374166/78244042-870bf380-74ed-11ea-999a-ac764d559818.png">

</details>

#### Changes
<!-- What are the changes made in this pull request? -->

- Add the MAC settings page
- Fix merging device parts in the js SDK for array values (before arrays were merged to an object `{ "0": value, "1": value }`
- Extend the `<KeyValueMap />` component to use indices as keys and render only the value input

#### Notes for Reviewers
<!--
NOTE: This section is optional.

Motivate briefly why it is implemented this way, if that deviates from the
implementation proposal in the referenced issues.
- How should your reviewers approach this pull request?
- @mention reviewers with special requests or questions for them
-->

1. @rvolosatovs suggested to omit the `RX_DELAY_0` value for the `rx1_delay` field, so you wont find it in the select (@johanstokking  is this fine?)

@rvolosatovs Please verify if works as expected

#### Checklist
<!-- Make sure that this pull request is complete. -->

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, database and configuration, according to the stability commitments in `README.md`.
- [x] Testing: The changes are covered with unit tests. The changes are tested manually as well.
- [ ] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

johanstokking commented:
>@rvolosatovs suggested to omit the `RX_DELAY_0` value for the `rx1_delay` field, so you wont find it in the select (@johanstokking  is this fine?)

Yes!

rvolosatovs commented on pkg/webui/console/views/device-mac-settings/device-mac-settings.js:
```diff
@@ -0,0 +1,244 @@
+  pingSlotPeriodicityValue: '{count, plural, one {every second} other {every {count} seconds}}',
+  resetsFCnt: 'Resets Frame Counters',
+  resetWarning: 'Reseting is insecure and makes your device susceptible for replay attacks',
+  rx1DelayTitle: 'Rx1 Delay',
```
```suggestion
  rx1DelayTitle: 'RX1 Delay',
```

rvolosatovs commented on pkg/webui/console/views/device-mac-settings/device-mac-settings.js:
```diff
@@ -0,0 +1,244 @@
+  resetsFCnt: 'Resets Frame Counters',
+  resetWarning: 'Reseting is insecure and makes your device susceptible for replay attacks',
+  rx1DelayTitle: 'Rx1 Delay',
+  rx1DelayDescription: 'Class A Rx1 delay. Rx2 delay is Rx1 delay + 1 second.',
```
```suggestion
  rx1DelayDescription: 'Class A RX1 delay. RX2 delay is RX1 delay + 1 second.',
```

More below

rvolosatovs commented on pkg/webui/console/views/device-mac-settings/validation-schema.js:
```diff
@@ -0,0 +1,114 @@
+    if (typeof freq !== 'undefined') {
+      return !isNaN(parseInt(freq))
+    }
+
```
```suggestion
```

rvolosatovs commented on pkg/webui/console/views/device-mac-settings/validation-schema.js:
```diff
@@ -0,0 +1,114 @@
+        }),
+      otherwise: schema => schema.strip(),
+    }),
+    ping_slot_periodicity: Yup.lazy(periodicity => {
```
If Class B is set for multicast devices, this must be set as well

bafonins replied:
`ping_slot_periodicity` field is shown and validated only for devices with `supports_class_b = true`. not sure what you mean here

rvolosatovs replied:
You should not be able to create multicast devices with `supports_class_b == true` if this value is unset.

bafonins replied:
You cannot create multicast devices via the Console atm

rvolosatovs replied:
Is there an issue already created about it?

rvolosatovs commented on pkg/webui/console/views/device-mac-settings/device-mac-settings.js:
```diff
@@ -0,0 +1,244 @@
+                component={Select}
+                options={pingSlotPeriodicityOptions}
+              />
+            )}
```
`mac_settings.ping_slot_frequency.value` must be available as well (otherwise class B devices cannot be created in `AU_915_928`, for example)

bafonins replied:
`ping_slot_periodicity` field is shown and validated only for devices with `supports_class_b = true`. not sure what you mean here

rvolosatovs replied:
Please reread my comment. I'm talking about the frequency here.

bafonins replied:
added

rvolosatovs commented on pkg/webui/console/views/device-mac-settings/device-mac-settings.js:
```diff
@@ -0,0 +1,244 @@
+                  component={Select}
+                  options={rx1DelayOptions}
+                />
+                <Form.Field
```
`mac_settings.rx1_data_rate_offset` should be available for both OTAA and ABP, but not for multicast

rvolosatovs commented on pkg/webui/console/views/device-mac-settings/validation-schema.js:
```diff
@@ -0,0 +1,114 @@
+      })
+    }),
+    rx1_data_rate_offset: Yup.number().when('$activation_mode', {
+      is: mode => mode === ACTIVATION_MODES.ABP,
```
Only ABP non-multicast

bafonins replied:
it is exactly how it works now

rvolosatovs commented on pkg/webui/console/views/device-mac-settings/validation-schema.js:
```diff
@@ -0,0 +1,114 @@
+      }
+
+      return Yup.object().when('$activation_mode', {
+        is: mode => mode === ACTIVATION_MODES.ABP,
```
Only ABP non-multicast

bafonins replied:
it is exactly how it works now

rvolosatovs commented on pkg/webui/console/views/device-mac-settings/device-mac-settings.js:
```diff
@@ -0,0 +1,244 @@
+            <Message component="h4" content={m.setMacSettings} />
+            {isABP && (
+              <>
+                <Form.Field
```
`mac_settings.rx1_delay` should be available for ABP, but not for multicast

bafonins replied:
it is exactly how it works now

bafonins commented:
Fixed issues and updated screenshots

> MAC settings must be available already during creation of device

Not in this PR

> Some errors seem not to be handled

Cant reproduce

> It should be possible to change "Supports class B" both on creation and update

What exactly should be added to the Console?

> It should be possible to change "Supports class C" on update

It is already possible in the general settings page in the network server section
<img width="613" alt="Screenshot 2020-04-02 at 14 26 23" src="https://user-images.githubusercontent.com/16374166/78244278-f2ee5c00-74ed-11ea-87cc-da903de912f6.png">

rvolosatovs commented on pkg/webui/console/views/device-mac-settings/device-mac-settings.js:
```diff
@@ -0,0 +1,272 @@
+                <Form.Field
+                  type="number"
+                  min={100000}
+                  step={1}
```
The step *SHOULD* be `100` Hz in fact

bafonins commented on pkg/webui/console/views/device-mac-settings/device-mac-settings.js:
```diff
@@ -0,0 +1,272 @@
+            {!isMulticast && (
+              <>
+                {isABP && (
+                  <Form.Field
+                    title={m.rx1DelayTitle}
+                    description={m.rx1DelayDescription}
+                    name="mac_settings.rx1_delay.value"
+                    component={Select}
+                    options={rx1DelayOptions}
+                  />
+                )}
+                {(isABP || isOTAA) && (
+                  <Form.Field
+                    title={m.rx1DataRateOffsetTitle}
+                    name="mac_settings.rx1_data_rate_offset"
+                    component={Select}
+                    options={dataRateOffsetOptions}
+                  />
```
> These should number type inputs as well with respective min/max props

https://github.com/TheThingsNetwork/lorawan-stack/pull/2278#pullrequestreview-386199884

> Make RX1 Data Rate Offset a select, same as other fields

kschiffer replied:
@rvolosatovs Why should it be?

rvolosatovs replied:
It does not need to be a select, perhaps the slider is even a better approach, the key here, however, is consistency, so RX1 data rate offset, RX2 data rate index and RX1 delay all should have the same method of selection.

kschiffer replied:
> It does not need to be a select, perhaps a slider is even a better approach, the key here, however, is consistency, so RX1 data rate offset, RX2 data rate index and RX1 delay all should have the same method of selection.

Right. So these should all be number type inputs with min/max constraints. A `select` is not an appropriate input type for selecting from a range of numbers.

bafonins commented on pkg/webui/console/views/device-mac-settings/device-mac-settings.js:
```diff
@@ -0,0 +1,272 @@
+                )}
+              </>
+            )}
+            <Form.Field
+              title={m.rx2DataRateIndexTitle}
+              description={m.rx2DataDateIndexDescription}
+              name="mac_settings.rx2_data_rate_index.value"
+              component={Select}
+              options={dataRateIndexOptions}
+            />
```
https://github.com/TheThingsNetwork/lorawan-stack/pull/2278#discussion_r404052093

bafonins commented on sdk/js/src/service/devices/index.js:
```diff
@@ -33,6 +33,7 @@ class Devices {
     if (!api) {
       throw new Error('Cannot initialize device service without api object.')
     }
+
```
will remove after rebase

bafonins commented:
> Thus, my suggestion here would be to make MAC Settings a tab under General Settings.

So a separate item or within the `Network Layer` section?

> I understand that these are separated on the grounds of the underlying end device data structure, but I would still suggest moving these fields into the `MAC Settings`.

I guess you meant 'I would still suggest moving these fields into the **General Settings**.

kschiffer commented:
> So a separate item or within the `Network Layer` section?

Separate item.

> I guess you meant 'I would still suggest moving these fields into the **General Settings**.

I mean moving the MAC/PHY selectors from the `Network Layer` to the `MAC Settings` section (within the `General Settings` page).

rvolosatovs commented:
> Not in this PR

Why not? I still have troubles figuring out why the "update" is separate from "create". None of the functionality added in the PR actually works for ABP devices, which do not reset FCnt. MAC settings are only applied on generation of a new MAC state of a device, which for OTAA happens on each join and for ABP happens only on creation and on FCnt reset for pre-1.1 devices. Regardless, it's a very bad UX that ABP device user's devices are broken on creation(because they still have to update the MAC settings and trigger a MAC state reset somehow).
I think we should reconsider merging this PR until MAC settings are available on creation (which IMO perfectly belongs in this PR), otherwise it's a half-baked feature, which IMO does not belong in `master`.

bafonins commented:
> I think we should reconsider merging this PR until MAC settings are available on creation (which IMO perfectly belongs in this PR), otherwise it's a half-baked feature, which does not belong in master.

Then, we will add this sometime after https://github.com/TheThingsNetwork/lorawan-stack/issues/579
